packages/@aws-cdk/aws-lambda/lib/function-hash.ts

```diff
@@ -9,15 +9,45 @@ export function calculateFunctionHash(fn: LambdaFunction) {
 
   // render the cloudformation resource from this function
   const config = stack.resolve((functionResource as any)._toCloudFormation());
+  // config is of the shape: { Resources: { LogicalId: { Type: 'Function', Properties: { ... } }}}
+  const resources = config.Resources;
+  const logicalId = Object.keys(resources)[0];
```

Comment:
Why is this change needed here? Feels unrelated. 

If the CfnFunction update resulted in hash changes and you think this is safe, we can easily update the hash without deploying by running `yarn cdk-assert --dry-run`.

We should follow up with a fix to how lambada calculates hash

Reply:
But reverting these changes means every single customer that updates to the new CDK version will get their Function's Versions replaced...?

Reply:
Why? Is this due to field order changes?

Reply:
Yes. The previous properties of the `CfnFunction` class result in `JSON.stringfy()` [from here](https://github.com/aws/aws-cdk/blob/0b6f0dd3838d6be45455aacb2c88472425302d52/packages/%40aws-cdk/aws-lambda/lib/function-hash.ts#L14) returning

```
{"Resources":{"Fn9270CBC0":{"Type":"AWS::Lambda::Function","Properties":{"Code":{"ZipFile":"foo"},"Role":{"Fn::GetAtt":["FnServiceRoleB9001A96","Arn"]},"Handler":"index.handler","Runtime":"nodejs10.x"}}}}
```

With the new properties (`ImageConfig` and `PackageType`) added to `CfnFunction`, the result is

```
{"Resources":{"Fn9270CBC0":{"Type":"AWS::Lambda::Function","Properties":{"Code":{"ZipFile":"foo"},"Handler":"index.handler","Role":{"Fn::GetAtt":["FnServiceRoleB9001A96","Arn"]},"Runtime":"nodejs10.x"}}}}
```

Which results in a different hash.

Reply:
To correct myself slightly, the addition of the `ImageConfig` and `PackageType` does not matter here. What changes the order is the fact that the `Handler` and `Runtime` properties have been made optional (from required) in this spec version, and our code generation takes [required properties before optional ones](https://github.com/aws/aws-cdk/blob/0b6f0dd3838d6be45455aacb2c88472425302d52/tools/cfn2ts/lib/codegen.ts#L163-L168).